File: rawDroid/src/main/java/com/anthonymandra/image/ImageConfiguration.java
package com.anthonymandra.image;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.anthonymandra.rawdroid.FullSettingsActivity;

public abstract class ImageConfiguration {
    public enum  ImageType
    {
        jpeg,
        tiff,
        raw
    }

    // For more than one config setting
    protected static final String DELIMITER = "_";

    public abstract ImageType getType();
    public abstract String getExtension();
    protected abstract void parse(String config);
    protected abstract String convertToPreference();

    public static ImageConfiguration loadPreference(Context c)
    {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(c);
        String typeString = pref.getString(FullSettingsActivity.KEY_DefaultSaveType, null);
        if (typeString == null)
            return null;
        String config = pref.getString(FullSettingsActivity.KEY_DefaultSaveConfig, null);

        ImageType type = Enum.valueOf(ImageType.class, typeString);
        switch(type)
        {
            case jpeg:
                JpegConfiguration jc = new JpegConfiguration();
                jc.parse(config);
                return jc;
            case tiff:
                TiffConfiguration tc = new TiffConfiguration();
                tc.parse(config);
                return tc;
            default:
                return null;
        }
    }

    public void savePreference(Context c)
    {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(c);
        SharedPreferences.Editor editor = pref.edit();
        editor.putString(FullSettingsActivity.KEY_DefaultSaveType, getType().toString());
        editor.putString(FullSettingsActivity.KEY_DefaultSaveConfig, convertToPreference());
        editor.apply();
    }
}

File: app/src/main/java/com/anthonymandra/image/ImageConfiguration.kt
package com.anthonymandra.image

import android.content.Context
import android.preference.PreferenceManager

import com.anthonymandra.rawdroid.settings.StorageSettingsFragment

abstract class ImageConfiguration {

	abstract val type: ImageType
	abstract val extension: String
	abstract val parameters: String

	enum class ImageType {
		JPEG,
		TIFF,
		RAW
	}

	fun savePreference(c: Context) {
		val pref = PreferenceManager.getDefaultSharedPreferences(c)
		val editor = pref.edit()
		editor.putString(StorageSettingsFragment.KEY_DefaultSaveType, type.toString())
		editor.putString(StorageSettingsFragment.KEY_DefaultSaveConfig, parameters)
		editor.apply()
	}

	companion object {

		fun from(type:ImageType, config: String?): ImageConfiguration {
			return when (type) {
				ImageType.JPEG -> if (config != null) JpegConfiguration(config) else JpegConfiguration()
				ImageType.TIFF -> if (config != null) TiffConfiguration(config) else TiffConfiguration()
				else -> if (config != null) JpegConfiguration(config) else JpegConfiguration()
			}
		}

		fun fromPreference(c: Context): ImageConfiguration? {
			val pref = PreferenceManager.getDefaultSharedPreferences(c)
			val typeString = pref.getString(StorageSettingsFragment.KEY_DefaultSaveType, null) ?: return null
			val config = pref.getString(StorageSettingsFragment.KEY_DefaultSaveConfig, null)

			val type = ImageType.valueOf(typeString)
			return from(type, config)
		}
	}
}
